### ISSUE TITLE
reload_check_config is racy/incomplete

### ISSUE THREAD

**Issue by wdauchy** [acassen/keepalived#1756]
we have a setup which often reload keepalived process (i.e. every 15 seconds at some periods of the day). Another in-house process is responsible to produce keepalived configurations files.
We sometimes faced several racy situations, two examples:

1. config file included within the main one was not present at the precise moment of the checker reload process.
Of course it should not happen but we know bugs happen, and will always happen.
- `reload_check_config` is parsing config files, everything is ok
- one of the config file included in the main one somehow removed from the filesystem
- checker process loads an incomplete config and removes most of the current setup.

2.  checker process was missing file descriptors quotas, so what happened was:
- `reload_check_config` is parsing config files, everything is ok
- checker process reloads, but cannot open the file, as it lacks file descriptors.

As in 1- this should not happen, and it is probably hiding another issue in checker process.

In both cases we absolutely don't want the checker process to continue, but better leave instead of continue reloading an empty config, and erase a critical config in place.
An easy safer workaround would be:
```
diff --git a/keepalived/check/check_daemon.c b/keepalived/check/check_daemon.c
index 5e03cc4c..cda85617 100644
--- a/keepalived/check/check_daemon.c
+++ b/keepalived/check/check_daemon.c
@@ -307,6 +307,8 @@ start_check(list_head_t *old_checkers_queue, data_t *prev_global_data)
        }
 
        init_data(conf_file, check_init_keywords);
+       if (reload && global_data->reload_check_config && get_config_status() != CONFIG_OK)
+               exit(KEEPALIVED_EXIT_CONFIG);
 
        if (reload)
                init_global_data(global_data, prev_global_data, true);
```

i.e. from our point of view it's better to stop when you detected a load issue than risky breaking your current config. I do not have any other issue for now than this as the current `reload_check_config` is based on forking an extra process.

I'm opening this issue to start a discussion but I'm ok to propose a possible implementation.

Thanks,

**Comment by pqarmitage** [comment]
@wdauchy I think there are two issues you have identified here:

1. What to do if a file referenced in an include statement is missing
        I think there are some wider issues with what you have described, and I will discuss them below.

2. The checker process running out of file descriptors.
I think there are two reasons that the checker process could run out of file descriptors.
    a) There is a bug and the checker process is not closing configuration files
    b) You have a huge depth of include files, and the checker process is simply running out of file descriptors, since when keepalived processes an `include` statement, it needs an extra file descriptor since it keeps the previous file open.
Please can you provide the following info:
i) what depth of `include`s you have
ii) the soft and hard limits for "Max open files" in /proc/PID/limits where PID is the pid of the checker process
iii) whether you use any wildcard characters in the include file names.
iv) When the checker process cannot open a file due to a lack of file descriptors, can you please provide the output of `lsof -p PID` and `cat /proc/PID/limits`, where PID is the pid of the checker process.

I have looked at a) above and keepalived does not appear to be leaking any file descriptors, including following a reload.

Now back to point 1. If I understand the situation correctly, I think what you are doing cannot work reliably. My understanding is that you have a process that updates the configuration files, and I am not sure whether it is that process or another process that signals keepalived to reload the configuration.

Any change in the configuration files while keepalived is reloading has the potential to cause a problem. If a config file has been deleted, if the include statement uses a wildcard to reference the include file, it is not possible to know that the file should be present but is actually missing; I think we can do something if the include statement does not use a wildcard though. I think the worse problem though is if a configuration file is written to/delete/created while keepalived is reloading. If that happens, then keepalived is almost certain not to load a configuration that you want. The only way that I think this can work is if your process(es) monitor the keepalived logs to ensure that keepalived has completed the reload, and until the reload has completed your process(es) must not modify any of the configuration files. Currently, the checker process writes "Reinitializing ipvs" after it has read all the configuration files following a reload, and writes "Initializing ipvs" after it has read the configuration files when it starts. Unfortunately there is no equivalent for the VRRP process. I am happy to add a new log message that writes "Reload complete", or "Startup complete" for all three processes (checker, VRRP and BFD) if that will help.

Your suggested code change of the checker process exiting if the config is not valid is not something we would add to the code. If the checker process exits, the old IPVS configuration will still be set; provided keepalived is not run with the `-R` option, the checker process will immediately be restarted, and another attempt will be made to load the configuration. The problem is that any real or virtual servers that were in the old configuration but are not in the new configuration will not be removed. Of course, if the configuration is not valid the keepalived checker process will repeatedly exit and be restarted.

**Comment by wdauchy** [comment]
Thank you @pqarmitage for this analysis. it seems like we are in case 2.a because we only include one file. For now I manually increased the number of FD available so we can understand whether there is a long-running leak which is hard to see (e.g servers being removed?).

For the reload itself I think we should think to rework the current `reload_check_config` option so that we can rely on the same `open` call and parsing and avoid relying on an extra process. It would remove:
- racy config parse (e.g the file is being updated between the two actions)
- detect fd limits in master process (even if your extra process has successfully parsed your file, it does not mean your process will be able to open the file)

outside of that I had more thoughts about making the reload safer: i.e avoid replacing a file while a reload is happening. The only sane solution I found is to make use of systemd `sd_notify` API; so the process can warn whether a reload is still ongoing. I made several implementation attempts but the current design of reload makes it very difficult to make the master process aware of "are other processes done with reload?". However resolving this issue would provide a sane API for external tools like ours to be able to assess whether it is ok to replace the current config file.
As a side note, that's what we use in haproxy to make sure we don't overlap two reloads.

**Comment by pqarmitage** [comment]
@wdauchy I have just merged two commits, 32e7804 and e6c582f that I think will have with some of the issues you outlined above.

Commit 32e7804 makes keepalived read the configuration file(s) only once. As it reads the file(s) it writes them to a memory based file, and all the other processes reading the configuration during the reload read it from the memory based files. This ensures that the configuration that is read by the running processes is exactly the same as the configuration that is read by the reload_check_config process. The same technique is also used 

Commit e6c582f adds synchronisation support between processes updating configuration files and keepalived reading those files. The `global_defs` option `reload_file` specifies a file that keepalived will create before starting to read the configuration files, if it doesn't exist, and then keepalived removes the file after it has finished reading the configuration files.

The best way to use this feature is as follows:
1. The process writing the new configuration writes the reload_file before it signals keepalived to reload (the default file is `keepalived.reload` in the PID file directory).
1. The process updating the configuration signals keepalived to reload the configuration (or `systemctl reload keepalived`) and stops updating configuration files.
1. The process updating configation files waits for the reload_file to be removed.
1. The process updating configuration files can now start updating configuration files again.

If at step 1 the reload_file is created with non-zero size, keepalived will clear the file when it starts reloading, so it is possible to detect when keepalived actually starts reloading.

**Comment by wdauchy** [comment]
Thank you @pqarmitage 
Commit 32e7804 looks very useful as it will prevent opening new file descriptors while reloading. It racy the racy part on file descriptors.

I'm a bit puzzled by commit e6c582f as we do not have a way to make sure the reload process is completely done in all init processes. We are facing very strange issues after reloads (process stuck for no obvious reason), and we would like to make sure to don't emit a new reload while, for example, the checker process is initialising all the new checks. That's why I was referring to the systemd API. Here this option only protects the racy read of the file, not the full reload process. Does that make sense?

**Comment by pqarmitage** [comment]
Regarding file descriptors, I am not aware of any race condition that exists with them. The motivation for commit 32e7804 was to ensure that we know that all the keepalived processes have read exactly the same configuration, and that it is the same configuration that was read by the process that checks the configuration before a reload.

I am not aware of any file descriptor problem, other than you have mentioned that you have one. The periodic output of the following script would be interesting:
```
#!/bin/bash

PID_DIR=${1:-/run}

for p in keepalived checkers vrrp bfd; do
	[[ ! -f ${PID_DIR}/$p.pid ]] && continue;

	echo; echo $p
	lsof -p $(cat ${PID_DIR}/$p.pid) 2>/dev/null | grep -v " mem "
done
```

Yes, commit e6c582f protects the race condition that exists if processes external to keepalived update keepalived's configuration files while keepalived is reading them; this is something that is essential if there is any risk that processes might update the config files while keepalived is reading them.

I understand from your previous message that you have a concern that reloading keepalived's configuration may cause a problem for new checkers that have not run since the configuration was last reloaded. Commit 47cc611d might help with that. I don't think there should be a problem with reloading keepalived before a checker has completed running for the first time. However, I need to think through the implications of reloading while a checker is running; particularly what happens if a checker is removed. I have not thought about this before.

I will have a look at the systemd sd_notify API and see if we can make integrate with that, although we need to ensure that keepalived can run reliably as a standalone program and also when using other init systems.

It would be very helpful to see a typical configuration that you use, and the sort of changes that you make, so that I can try some tests.

**Comment by pqarmitage** [comment]
@wdauchy Are you using any SSL_GET checkers? There appears to be the potential there to leak a file descriptor on reload.

**Comment by wdauchy** [comment]
no we do not use SSL_GET checkers; however we were also thinking about the case when a checker is removed and a reload is triggered.this might be one of the racy situation.

sd_notify should not be a problem as the main purpose is to warn systemd another reload is still ongoing. a user can issue a `systemcl is-active` to get an idea whether the process is still reloading. The difficulty in the current design is to know when all processes are done wth reload.

I will send you my config and the output of `lsof` privately to avoid leaking too much public information though.

**Comment by pqarmitage** [comment]
There certainly used to be a problem of file descriptors being leaked during a reload (see #1177), which commit 7dfad19 was intended to resolve. I noticed earlier that there are a couple of instances of `thread_add_read()` which pass the close_on_reload parameter as `false`, whereas it should be `true`. I also need to check whether there is a potential memory leak, both for SSL_GET and HTTP_GET relating to a reload, in respect of the checker->req structure.

**Comment by wdauchy** [comment]
commit 7dfad198a706b5616adfd28e3861d3b2b4572dff seems to be already on our version as we are on v2.1.5

Thanks for the investigation around around `close_on_reload`

**Comment by pqarmitage** [comment]
@wdauchy I have looked at the `lsof`output and can see that you have 17 sockets in CLOSE_WAIT state. Does the number of sockets in CLOSE_WAIT state gradually increase over time?

Since the FDs are in the range 8 to 70, I think you must have had at least 63 HTTP checkers running simultaneously, which is not unreasonable given that you have nearly 5000 checkers configured. It is interesting that the "DEVICE" number for 13 of the sockets in the CLOSE_WAIT state are in a small numerical range (fds 39 to 70), which suggest that they all got "lost" at the same time, and at a time when there were at least 63 checkers running.

When you have a keepalived process with sockets left in the CLOSE_WAIT state, it would be very interesting to see the state of the internal thread scheduler in keepalived. To achieve this, can you build keepalived with the `--enable-dump-threads` configure option. Then when you have a number of sockets in the CLOSE_WAIT state, run (as root) `kill -$(keepalived/keepalived --signum=TDUMP) $(cat /run/keepalived/checkers_keepalived-lvs.pid)`. That should write the thread information to syslog; there is a risk that syslog might discard some of it due to too high a logging rate, but we might get some useful information. If you could send the syslog output and the matching `lsof` that might give a further indication about what is happening.

**Comment by wdauchy** [comment]
the close_wait state seems a good lead indeed. It's hard to see whether it is a leak, but if I look at our graph, I can see that it is indeed increasing very slowly (as I was expecting, because we trigger a FD limit after a while), up to 21 on the machine I looked, and coming back to zero on a restart only.

I will try to have the dump threads build soon though. @pierrecdn could you take over this issue? :)

**Comment by pierrecdn** [comment]
Sure I can, I will just have a small lag during a few days (coming back from vacations). Will try to deploy this soon and keep you aware of the output.

**Comment by pqarmitage** [comment]
@wdauchy @pierrecdn I can now reproduce this problem, so there is no need for you to build a debug version of keepalived and run a test (of course if you are interested in doing so then I would welcome your feedback).

I am finding that when keepalived starts up, the checker process runs for about 75 seconds consuming near 100% CPU of 1 core, and during this time it is logging `Gained quorum 1+0=1 <= 130 for VS ...` and the time it takes loading the 48th virtual server is 26 seconds. Is this what you experience?

What I have done to reproduce the problem is to modify a network echo server (tcp_server.c in the test directory) so that it can respond to http requests, so it mimics a web server for the HTTP_GET checkers. I have set the "pseudo" web server to delay one second before sending the response to the HTTP GET, which means that the TCP sockets are all open for sufficiently long that a reload during the time that the HTTP_GET checkers have TCP connections open then triggers the "lost" file descriptors to occur. Running this way, the first reload I tried during the time that checkers were running caused 30 file descriptors to be "lost".

Fortunately, when I ran the test, I was running with full thread debugging enabled (it produced a 1.5Gb log file in under 5 minutes). What I can see is that at the same time as the reload signal was received by the checker process via a signalfd, 200 other FDs were returned as ready by epoll, these all being FDs for the HTTP_GET checkers. As keepalived steps through each of the FDs returned by epoll, it eventually gets to the signalfd, sees that it has a reload signal, and commences the reload. At that point there were still 30 other FDs that had been returned from epoll() that had not been processed; these are precisely the 30 FDs that were lost during the reload.

I can think of a relatively easy work around to stop the problem occurring, and will produce a patch for this later. I will however need to investigate why the 30 remaining FDs were "lost", since there may be some other areas that could suffer from the same problem.

**Comment by wdauchy** [comment]
thanks @pqarmitage for the investigations!

**Comment by pqarmitage** [comment]
@wdauchy Commit ba3ce49 resolves the file descriptor leak problem that I have found, and it is something that I would expect you to experience, although as you say only occasionally. If file descriptors relating to HTTP_GETs were returned by epoll_wait() at the same time as the signal_fd file descriptor was returned (this fd would be returned when keepalived was signalled to reload), then any file descriptors in the list returned by epoll_wait() that were after the signal_fd in the list would be "lost". This was down to a simple bug in the code, and was quite obvious once identified.

With the config that you sent to me, I would expect there to be 75 "Gained quorum" messages when keepalived is starting up, but generally not otherwise unless HTTP_GETs failed and then were successful again.

I think the problems that you have identified in this issue are:
1. The keepalived configuration can change between the time when the check_config process runs and when the main processes then read the configuration themselves
1. You are experiencing occasional file descriptor leaks occurring during reloads which were accumulating over time and eventually caused the keepalived process to run out of available file descriptors, causing a failure to open configuration files during a reload.
1. There was a suspicion that reloading keepalived before all the HTTP_GETs had completed for the first time was a cause of/contributed to the file descriptor leaks.
1. There was no mechanism to know when keepalived had completed the reload process, which is needed so that configuration files could again safely be updated and keepalived signalled to reload
1. The configuration parser did not report configuration errors if `include` files were missing
1. Using the sd_notify API would be beneficial for reloading

The position I think we have now reached is:
1. Commit 3247f6b0 resolves this point, since the configuration files are now read only once, and are copied to a temporary file which the other processes read.
1. Commit ba3ce496 resolves the file descriptor leak that I could reproduce during reloads. This is something you are likely to have experienced occasionally, and so I believe that your file descriptor leak problem should now be resolved.
1. There is no issue reloading keepalived before the HTTP_GET checks have completed for the first time, although point 2 above made it look as though there was.
1. Commit e6c582f1 provides a mechanism for knowing when keepalived has completed reading the configuration files while reloading, thereby allowing automated processes to initiate a reload and then detect when it is safe to start updating the configuration files again.
1. You have provided a patch, pull request #1755 that detects when an include file is missing, and causes the config_check process to fail if that is the case. I will merge that later, and also some commits that extend your approach so that other users should be able to benefit from your update in other ways. However, if you are able to use the new mechanism outlined in point 4. above, this issue of a missing include file should not occur; if you can't use the approach in point 4. above, then I think there is still an issue that an update process may start rewriting the configuration file while keepalived is reading it; it won't be missing, but it may be invalid, or worse still valid but not what you want.
1. I will look further into using the sd_notify API. I wasn't previously aware of it, and I do like the idea.

Many thanks for all your help with this. You have certainly brought about the fixing of one bug in keepalived, and the improvement of the reading of configuration files and the reload process.

If I have missed any points, or you thinks there are still some problems that need resolving, please don't hesitate to update this issue with the details.

**Comment by wdauchy** [comment]
Thank you very much @pqarmitage, I think your last comment covers everything. Glad that it helped improving keepalived.

I've backported commit ba3ce49 so we can test on top of v2.1.5 as it applies cleanly without any change. For the rest I think it is safer to wait for the next release.

**Comment by wdauchy** [comment]
just a small followup, first observations seemed to show commit ba3ce49 was fixing the leak we observed. @pierrecdn will double confirm next week.

**Comment by pierrecdn** [comment]
I confirm we did not see any leak anymore (and this since 4 days now).

Maybe this issue can be closed now? The other points are tracked in #1755, and we can open something else to integrate with sd_notify?

**Comment by pqarmitage** [comment]
@pierrecdn Many thanks for confirming that the file descriptor leak appears to be resolved now.

I am easy whether this issue remains open to deal with sd_notify and the workings of `include`, or whether new issues are opened.

**Comment by wdauchy** [comment]
ok lets keep that one for the remaining tasks

**Comment by pqarmitage** [comment]
I have now merged pull request #1755. I have also added some updates, and changed the keyword `glob_strict` to `include_check`, since I think that is a more meaningful name; many apologies for changing my mind on the keyword.

`include_chek` now supports a number of options for enabling and disabling different include checks. There are also new include variants `includer`, `includem`, `includew`, `includeb` and `includea`. In addition to this I have added a new command line option `-e` that will make keepalived terminate if that are any include file errors; the purpose of this is so that you can stop keepalived being able to load if the configuration is invalid for a reload.

Finally, I have added systemd notification support. Unlike haproxy, keepalived works out for itself if it being run by keepalived and the service has `type=notify`. If keepalived detects this, then it will notify systemd tht startup has completed, that it is reloading, reloading has completed, and that it is shutting down.

I think that completes all the features you have suggested. Please update this issue, or open a new issue, if there are any points outstanding or this doesn't satisfy your needs.

**Comment by wdauchy** [comment]
Thank you @pqarmitage !
for systemd notifications I was curious how you would implement it, I saw the `child_reloaded` function, good point!
